File: Go/sortingAlgorithms.go
package main

import (
	"fmt"
	"time"
)

func main() {
	fmt.Printf("Sorting Algorithms\n------------------\nImplemenation of different sorting algorithms all sorting the same orignial unsorted array. \nEach implementation will be timed to give an idea of how fast each algorithm is/ can be.\n")
	unsortedArray := []int{4, -34, 23, 50, 0, -7}
	fmt.Printf("\nUnsorted slice: %v\n\n", unsortedArray)

	// Bubble sort using minimum
	bsStart := time.Now()
	bsArr := bubbleSortMin(unsortedArray)
	fmt.Printf("Bubble sort (Min.): \n\t> Input: %v\n\t> Result: %v\n\t> Time: %v\n", unsortedArray, bsArr, time.Since(bsStart))

	// Selection sort using minimum
	ssStart := time.Now()
	ssArr := selectionSort(unsortedArray)
	fmt.Printf("Selection sort (Min.): \n\t> Input: %v\n\t> Result: %v\n\t> Time: %v\n", unsortedArray, ssArr, time.Since(ssStart))

	// Merge sort (ex-situ variant)
	msesStart := time.Now()
	msesArr := mergeSortExSitu(unsortedArray)
	fmt.Printf("Merge sort (ex-situ): \n\t> Input: %v\n\t> Result: %v\n\t> Time: %v\n", unsortedArray, msesArr, time.Since(msesStart))
}

func swap(a, b *int) {
	temp := *a
	*a = *b
	*b = temp
}

func bubbleSortMin(unsorted []int) []int {
	// Copy slice to leave original array untouched
	sorted := make([]int, len(unsorted))
	copy(sorted, unsorted)

	// Sort
	for i := 0; i < len(sorted); i++ {
		for j := len(unsorted) - 2; j >= i; j-- {
			if sorted[j] > sorted[j+1] {
				swap(&sorted[j], &sorted[j+1])
			}
		}
	}
	return sorted
}

// Selection sort algorithms selecting the minimum each iteration.
// Runtime complexity: BC = WC = AC = O(n²)
func selectionSort(unsorted []int) []int {
	// Copy slice to leave original array untouched
	sorted := make([]int, len(unsorted))
	copy(sorted, unsorted)

	// Sort
	for i := 0; i < len(sorted); i++ {
		min := i
		for j := i; j < len(sorted); j++ {
			if sorted[j] < sorted[min] {
				min = j
			}
		}
		swap(&sorted[i], &sorted[min])
	}
	return sorted
}

// Merge sort algorithm. This is the typical ex-situ implementation which has ideal runtime complexity.
// Runtime complexity: O(n*log(n))
func mergeSortExSitu(unsorted []int) []int {
	// Copy slice to leave original array untouched
	sorted := make([]int, len(unsorted))
	copy(sorted, unsorted)
	return mergeSortExSituInternal(sorted, 0, len(sorted)-1)
}

// INTERNAL: function called recursively to divide array into two halfs and merge those two halfs back together as required for merge sort.
func mergeSortExSituInternal(unsorted []int, first, last int) []int {
	if first < last {
		middle := first + ((last - first) >> 1)
		// sort left half
		mergeSortExSituInternal(unsorted, first, middle)
		// sort right half
		mergeSortExSituInternal(unsorted, middle+1, last)
		// merge
		mergeExSitu(unsorted, first, middle, last)
	}
	return unsorted
}

// Merges values of left and right half of slice which are given by [first:middle+1] [middle+1:last+1]
// This is an ex-situ implementation which gurantees a runtime complexity of O(n) which is not possible with in-situ version
// Memory complexity: O(n)
// Runtime complexity: O(n)
func mergeExSitu(unsorted []int, first, middle, last int) []int {
	// length of current interval
	invalLen := last - first + 1
	// Ex-situ algorithms -> requires additional storage O(n) to run in O(n*log(n)) in BC/ AC/ WC
	sorted := make([]int, invalLen)
	// Define indices pointing to current element in left and right half of array
	left, right := first, middle+1

	// Merge values together in slice "sorted"
	for i := 0; i < invalLen; i++ {
		if left <= middle {
			if right <= last {
				if unsorted[left] <= unsorted[right] {
					sorted[i] = unsorted[left]
					left++
				} else {
					sorted[i] = unsorted[right]
					right++
				}
			} else {
				sorted[i] = unsorted[left]
				left++
			}
		} else {
			sorted[i] = unsorted[right]
			right++
		}
	}
	// copy values back to original array
	for i := 0; i < invalLen; i++ {
		unsorted[first+i] = sorted[i]
	}

	return unsorted
}

File: Python/sorting_algorithms.py
import random
from timeit import default_timer


def init():
    arrayToSortLecture = [34, 45, 12, 34, 23, 18, 38, 17, 43, 7]
    randArray = fill_array_rand(10, -100, 100)
    test_arr = [21, -6, -46, -14, 9, 43, -18, 9, 21, 20]
    arrayToSort = randArray

    print(f"""
    +-------------------------------------------+
    | Sorting Algorithms and their runtime      |
    +-------------------------------------------+
    Unsorted array: {arrayToSort}
                        
    """)

    # Insertion sort
    is_arr = arrayToSort.copy()
    is_start = default_timer()
    insertion_sort(is_arr)
    is_end = default_timer()
    print_result("Insertion sort", is_end - is_start, is_arr)

    # Selection sort
    ss_arr = arrayToSort.copy()
    ss_start = default_timer()
    selection_sort(ss_arr)
    ss_end = default_timer()
    print_result("Selection sort", ss_end - ss_start, ss_arr)

    # Bubble sort
    bs_arr = arrayToSort.copy()
    bs_start = default_timer()
    bubble_sort(bs_arr)
    bs_end = default_timer()
    print_result("Bubble sort", bs_end - bs_start, bs_arr)

    # Quick sort
    qs_arr = arrayToSort.copy()
    qs_start = default_timer()
    quick_sort(qs_arr, 0, len(qs_arr) - 1)
    qs_end = default_timer()
    print_result("Quick sort", qs_end - qs_start, qs_arr)


def fill_array_rand(size, randStart, randEnd):
    """
    Create an array with certain size filled with random values within a specified range.

    :param size: size of array
    :type size: int
    :param randStart: start of random value range
    :type randStart: int
    :param randEnd: end of random value range
    :type randEnd: int
    :returns: array of size with values between randStart and randEnd
    """
    return [random.randint(randStart, randEnd) for _ in range(size)]


def swap(array, idx_x, idx_y):
    """
    Swap two values within an array. Mutates array!

    :param array: array
    :type array: list[any]
    :param idx_x: index of a value
    :type idx_x: int
    :param idx_y: index of another value
    :type idx_y: int
    :return: array with swapped elements
    """
    temp = array[idx_x]
    array[idx_x] = array[idx_y]
    array[idx_y] = temp
    return array


def print_result(alg_name, time_elapsed, sorted_arr):
    """
    Print result of a sorting algorithm.

    :param alg_name: algorithm name
    :param time_elapsed: time elapsed in seconds (timeit.default_timer())
    :param sorted_arr: sorted array
    :return:
    """
    time_elapsed_ms = time_elapsed * 1000
    print(f"""
    > {alg_name}:
    \t Time elapsed: {time_elapsed_ms:.2f}ms\t Result: {sorted_arr}
    """)


def insertion_sort(array):
    """
    Insertion sort algorithm sorting in ascending order.
    Runtime:
    - Best Case: O(n) -> All elements are already in sorted order
    - Average Case: O(n^2) -> Simplification: values of array are evenly distributed (while loop will run approx. half as many times as in worst case but still because of gaussian summation formula O(n^2))
    - Worst Case: O(n^2)    -> All elements are sorted in descending order (while loop will run "to full extend" -> gaussian summation formula -> O(n^2))

    :param array: array to be sorted
    :type array: list[int]
    :return: sorted array
    """
    step_counter = 0
    swap_counter = 0
    if len(array) == 1:
        return array

    for j in range(1, len(array)):
        key = array[j]
        i = j - 1
        step_counter += 2
        while i >= 0 and array[i] > key:
            array[i + 1] = array[i]
            swap_counter += 1
            i -= 1
            step_counter += 4 # two comparisms and two assignments
        array[i + 1] = key
        step_counter += 3 # two comparisms and assignment
    print(f"Steps: {step_counter}, Swaps: {swap_counter}")
    return array


def selection_sort(array):
    """
    Selection sort algorithm sorting in ascending order.
    Here: Move minimum to front
    Alternative: Move maximum to back.
    Runtime: Average Case = Best Case = Worst Case = O(n^2) (gaussian summation formula!)

    :param array: array to be sorted
    :type array: list[int]
    :return: sorted array
    """

    for j in range(len(array)):
        min_idx = j
        for i in range(j, len(array)):
            if array[i] < array[min_idx]:
                min_idx = i
        swap(array, j, min_idx)
    return array


def bubble_sort(array):
    """
    Bubble sort algorithm sorting in ascending order.
    Here: Move mininum from back to front
    Alternative: Move maximum from front to back
    Runtime: Average Case = Best Case = Worst Case = O(n^2) (gaussian summation formula!)

    :param array: array to be sorted
    :type array: list[int]
    :return: sorted array
    """
    for j in range(len(array)):
        for i in range(len(array) - 2, j-1, -1):
            if array[i] > array[i+1]:
                swap(array, i, i+1)
    return array


def quick_sort(array, first, last):
    if first < last:
        pivot_elem = prepare_partition(array, first, last)
        print(array)
        print("Pivot:", pivot_elem, array[pivot_elem])
        quick_sort(array, first, pivot_elem - 1)
        quick_sort(array, pivot_elem + 1, last)
    return array


def prepare_partition(array, first, last):
    # Choose random pivot element and move it to front of array
    pivot_idx = first # random.randint(first, last)
    swap(array, pivot_idx, first)
    part = first - 1
    pivot = array[first]

    for i in range(first, last+1):
        if array[i] <= pivot:
            part += 1
            swap(array, part, i)

    # Swap pivot back to where it belongs
    swap(array, first, part)

    # return index of pivot element (left from it --> everything is lower, right from it -> everything is greater)
    return part

if __name__ == '__main__':
    init()
